### PR TITLE
MM-34389: Reliable websockets: First commit

### PR DESCRIPTION
**Pull request by agnivade**
This is the first commit which makes some basic changes
to get it ready for the actual implementation.

Changes include:
- A config field to conditionally enable it.
- Refactoring the WriteMessage along with setting the deadline
to a separate method.

The basic idea is that the client sends the connection_id
and sequence_number either during the handshake (via query params),
or during challenge_auth (via added parameters in the map).

If the conn_id is empty, then we create a new one and set it.
Otherwise, we get the queues from a connection manager (TBD)
and attach them to WebConn.

```release-note
NONE
```

https://mattermost.atlassian.net/browse/MM-34389

### CODE REVIEW

File: api4/websocket.go [hunk 1]
```diff
@@ -31,6 +38,46 @@ func connectWebSocket(c *Context, w http.ResponseWriter, r *http.Request) {
...
+				wc.WebSocket.Close()
+				return
+			}
+			wc.Sequence = int64(seq)
+			// Now if there have been past entries to be back-filled, we do it.
+			// First we find the right sequence number point.
+			// We start consuming from dead queue first, and then move to active queue
+		}
+		// In case of fresh connection id, sequence number is already zero.
+		wc.SetConnectionID(connID)
```
Comment (streamer45) [inline api4/websocket.go:78]: Should we do some basic validation before setting these values? I am thinking the sequence should be a positive number and the ID should be our usual 26 alphanumeric string.

Comment (agnivade) [reply api4/websocket.go:78]: Makes sense. :+1:

**Review by streamer45 — APPROVED** [review]
Looks good :+1:

File: api4/websocket.go [hunk 2]
```diff
@@ -62,7 +67,7 @@ func connectWebSocket(c *Context, w http.ResponseWriter, r *http.Request) {
 				return
 			}
 			seq, err := strconv.Atoi(seqVal)
-			if err != nil {
+			if err != nil || seq < 0 {
```
[the remark this answers is not in the document]
Comment (agnivade) [reply api4/websocket.go:70]: Nothing happens, it just skips the field.

```go
func NamedError(key string, err error) Field {
	if err == nil {
		return Skip()
	}
	return Field{Key: key, Type: zapcore.ErrorType, Interface: err}
}
```

File: services/telemetry/telemetry.go [hunk 3]
```diff
@@ -440,6 +440,7 @@ func (ts *TelemetryService) trackConfig() {
 		"thread_auto_follow":                                      *cfg.ServiceSettings.ThreadAutoFollow,
 		"enable_link_previews":                                    *cfg.ServiceSettings.EnableLinkPreviews,
 		"enable_file_search":                                      *cfg.ServiceSettings.EnableFileSearch,
+		"enable_reliable_websockets":                              *cfg.ServiceSettings.EnableReliableWebSockets,
```
Comment (isacikgoz) [inline services/telemetry/telemetry.go:443]: Hmm, not sure if we should start to collect this telemetry immediately.

**Review by isacikgoz — APPROVED** [review]
Looks good to me 👍 just a small thing.

File: services/telemetry/telemetry.go (hunk 3 above)
Comment (agnivade) [reply services/telemetry/telemetry.go:443]: Yeah, agree.

**Comment by agnivade** [comment]
No QA needed as the feature is not yet ready to be tested.

**Comment by amyblais** [comment]
@cwarnermm This adds a new config setting, was it discussed with Agniva that this doesn't need docs?

**Comment by cwarnermm** [comment]
@amyblais - Yes. @agnivade confirmed that this feature isn't ready to be used yet. He'll add release notes to the final PR when it can be used.

**Comment by agnivade** [comment]
Yep, confirmed. No need to advertise this just yet.

**Comment by agnivade** [comment]
@amyblais - This is good to be advertised for 5.36. I have let @cwarnermm know the text to add in the release notes.

**Pull request merged**